_posts/2020-11-30-flexible-and-safe-tuples-in-scala3.md

```diff
@@ -0,0 +1,387 @@
+of the elements of the tuple.
+That is:
+
+```scala
+(1, "2", 3.0)
+val res0: List[Int | String | Double] = (1, "2", 3.0).toList
+```
+This is interesting because the type information is somehow maintained.
+We can iterate over `list` and use pattern matching to apply the
+correct transformation, knowing exactly how many and what cases to
+treat:
+
+```scala
+// The compiler tells he cannot help with checking:
+// Non-exhaustive match
+(1, "2").toArray.map {
+  case i: Int => (i * 2).toString
+  case j: String => j
+}
+
+// The code compiles without errors or warning
+// the compile verified that we handled all possible cases
+(1, "2").toList.map {
+  case i: Int => (i + 2).toString
+  case j: String => j
+}
+```
+Read more about [Union types here](https://dotty.epfl.ch/docs/reference/new-types/union-types.html).
+
+We can also transform a tuple by appling a function to each element.
+The method, similarly to what we are used to with collections, is called
+`map`. The difference from collections (or functors) is however
+that they expect a `f: A => B` where `A` is the type of the elements
+of the collection.
+With tuples each element has a different type!
+How can we generalize the concept of a function whose argument type is
+not fixed ?
+We can use a **`PolyFunction`**. This is a more advanced syntax:
+
+```scala
+(1, 'a', "dog", 3.0).map[[X] =>> Option[String]]([T] => (t: T) => t match {
+  case i: String => i.toUpperCase
+  case j: Int => j.toString
+})
+```
+
+## Tuples can compose
+
+The last operator allows to pair the elements of two tuples.
+You might have guessed, it is called `zip`. If the two tuples have
+different lengths, the extra elements of the longest will be
+ignored:
+
+```scala
+val numbers = (1, 2, 3, 4, 5)
+val letters = ('a', 'b', 'c')
+
+numbers.zip(letters) // ((1, 'a'), (2, 'b'), (3, 'c'))
+```
+
+# Under the hood: new type operators of Scala 3
+
+I believe that the core new features that allows such a flexible
+implementation of tuples are **match types**.
+I invite you to read more about them [here](http://dotty.epfl.ch/docs/reference/new-types/match-types.html).
+
+Let's see how we can implement the `++` operator using this powerful
+construct. We will naively call our version `concat`
+
+DISCLAIMER: This section is a bit more advanced !
+
+## Defining tuples
+
+First let's define our own tuple:
+
+```scala
+enum Tup:
+  case EmpT
+  case TCons[H, T <: Tup](head: H, tail: T)
+```
+
+That is a tuple is either empty, or an element `head` which precedes
+another tuple. Using this recursive definition we can create
+a tuple as follow:
+
+```scala
+import Tup._
+
+val myTup = TCons(1, TCons(2,  EmpT))
+```
+It is not very pretty, but it can be easily adapted to provide
+the same ease of use as the previous examples.
+To do so we can use another exciting new Scala 3 feature: [extension methods](http://dotty.epfl.ch/docs/reference/contextual/extension-methods.html)
```

Comment:
Couldn't I have done the same thing in Scala 2 with an `implicit class`? It's fine to call attention to the nice new syntax, but as written, it leads the reader to believe this wouldn't have even been possible in Scala 2.

If it's just a syntax difference, it isn't "exciting" in this context.

Reply:
I am already using enums and match types, I did not think it would be necessary to mention `implicit`

Reply:
(I think simply dropping the words "exciting new" would address my concern. Anyway, no big deal.)